src/render/tutorials/api-gateway-security-jwt/2_0.html.md

```diff
@@ -0,0 +1,329 @@
+
+![Generating JWT using jwt.io](/img/blog/api-gateway-security/security-generate-jwt.png)
+
+1. Select `HS256` as the algorithm.
+2. Fill in the key you used in the configuration.
+3. Copy the generated token.
+
+Now let's try this `curl` command with `Authorization` header:
+
+```
+$ curl -X GET http://localhost:8092/api/secure/jwt -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIn0.vPWK59pl5GWimz8UVbL3CmrceSfmNvvCgyzwLVV9jT8'
+{"message":"Hello World From Knot.x with JWT!"}
+```
+
+## Functional tests
+
+It's always a good practice to test our code. Let's implement some tests that will automatically do the same things we did manually above.
+
+The tests we're about to write will be functional tests, that will be executed on a real, running Docker image, just after the `healthcheck` (described in [this tutorial](http://knotx.io/tutorials/getting-started-with-docker/2_0/)) executes successfully.
+
+Let's add a new file: `functional/src/test/java/com/project/test/functional/JWTAuthITCase.java`
+
+```java
+class JWTAuthITCase {
+
+  @Test
+  @DisplayName("GIVEN no authorization WHEN call JWT secured API EXPECT Unauthorized")
+  void givenNoAuthorizationWhenCallBasicAuthApiExpectUnauthorized() {
+    given()
+        .port(8092)
+      .when()
+        .get("/api/secure/jwt")
+      .then()
+        .assertThat()
+        .statusCode(401);
+  }
+
+  @Test
+  @DisplayName("GIVEN authorization WHEN call JWT secured API EXPECT Ok")
+  void givenAuthorizationWhenCallBasicAuthApiExpectOk() {
+    given()
+        .port(8092)
+        .header("Authorization", "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIn0.vPWK59pl5GWimz8UVbL3CmrceSfmNvvCgyzwLVV9jT8")
+      .when()
+        .get("/api/secure/jwt")
+      .then()
+        .assertThat()
+        .statusCode(200);
+  }
+}
+```
+
+We've created two simple tests using [REST-assured](http://rest-assured.io/). They do the same actions we did above:
+
+- call secured endpoint without authorization and expect `401 - Unauthorized` response
+- call the same endpoint with authorization and expect `200 - OK` response
+
+If you try rebuilding the project now (remember to stop any running containers before that) you'll notice that those test will be invoked. Now we're confident that if our future changes will somehow disable or modify our `/api/secure/jwt` logic, we'll know at build time!
+
+# Summary
+
+In this tutorial we have successfully secured our API Gateway with one of the most common authentication methods.
+
+The only thing left to do here for a complete application is to implement login functionality. After successful login, the token should be stored by the browser's cookie in a manner that the token is appended to every request, but is not available for the front-end logic. Lastly, our apache client should rewrite this cookie to a header before passing a request to Knot.x.
```

Comment:
Maybe we should name it more generic, like "proxy" (and say as above that we use for it e.g. Apache or Nginx).